### PR TITLE
DAOS-14260 dfs: add API to readdir with a filter

### PR DESCRIPTION
**Pull request by mchaarawi**
This allows user to utilize the pipeline API to execute a server side find.

Test-tag: test_daos_dfs_unit
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/client/dfs/dfs.c [hunk 1]
```diff
@@ -7446,3 +7442,405 @@ dfs_obj_fix_type(dfs_t *dfs, dfs_obj_t *parent, const char *name)
...
+#define EQ_F		"DAOS_FILTER_FUNC_EQ"
+#define BA_F		"DAOS_FILTER_FUNC_BITAND"
+#define AND_F		"DAOS_FILTER_FUNC_AND"
+#define OR_F		"DAOS_FILTER_FUNC_OR"
+#define COND_F		"DAOS_FILTER_CONDITION"
+
+int
+dfs_pipeline_create(dfs_t *dfs, dfs_predicate_t pred, uint64_t flags, dfs_pipeline_t **_dpipe)
+{
+	daos_size_t	bin_flen = strlen(BINARY_F);
```
[the remark this answers is not in the document]
Comment (mchaarawi) [reply src/client/dfs/dfs.c:7522]: sure that works too, but what would be the difference in this case?

File: src/client/dfs/dfs.c [hunk 2]
```diff
@@ -7446,3 +7442,405 @@ dfs_obj_fix_type(dfs_t *dfs, dfs_obj_t *parent, const char *name)
...
+
+	if (oids) {
+		iod.iod_nr ++;
+		recxs[2].rx_idx	= OID_IDX;
+		recxs[2].rx_nr	= sizeof(daos_obj_id_t);
+		record_len += recxs[2].rx_nr;
+	}
+	if (csize) {
+		iod.iod_nr ++;
+		recxs[3].rx_idx	= CSIZE_IDX;
```
Comment (wiliamhuang) [inline src/client/dfs/dfs.c:7743]: question: could "oids" be NULL? If "oids" is NULL, shall we still use "recxs[3]" or "recxs[iod.iod_nr - 1]" in these three lines?

Comment (mchaarawi) [reply src/client/dfs/dfs.c:7743]: yes indeed it can be null since it's optional. ill fix that, thx.

File: src/client/dfs/dfs.c [hunk 3]
```diff
@@ -7446,3 +7442,405 @@ dfs_obj_fix_type(dfs_t *dfs, dfs_obj_t *parent, const char *name)
...
+		d_iov_set(&dpipe->dkey_ft.data_type, BINARY_F, bin_flen);
+		dpipe->dkey_ft.data_len = DFS_MAX_NAME;
+
+		d_iov_set(&dpipe->const0_ft.part_type, CONST_F, const_flen);
+		d_iov_set(&dpipe->const0_ft.data_type, BINARY_F, bin_flen);
+		dpipe->const0_ft.num_constants = 1;
+		dpipe->const0_ft.constant = &dpipe->dkey_iov;
+		d_iov_set(dpipe->const0_ft.constant, dpipe->pred.dp_name, name_len);
+
+		d_iov_set(&dpipe->like_ft.part_type, LIKE_F, like_flen);
```
Comment (jolivier23) [inline src/client/dfs/dfs.c:7563]: so does a regular expression have to use raw string type (e.g TYPE_BINARY)?

**Review by jolivier23 — DISMISSED** [review]

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (mchaarawi) [reply src/client/dfs/dfs.c:7563]: i believe so yes. this pipeline API could use some re-work to make it user-friendly.

**Review by wiliamhuang — APPROVED** [review]

File: src/client/dfs/dfs.c (hunk 3 above)
Comment (jolivier23) [reply src/client/dfs/dfs.c:7563]: Would be simple enough to add another enum entry with the same value RAW_STRING=BINARY

**Review by jolivier23 — APPROVED** [review]

**Pull request merged**